fixed #10513 ContentSourceInputStream close with available

gregw opened the pull request:
address close issue discovered in  #10513 ContentSourceInputStream close by making it consume available content whilst seeking EOF

gregw committed:
fixed #10513 ContentSourceInputStream close with available

fixed #10513 ContentSourceInputStream close by making it consume available content whilst seeking EOF

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceInputStream.java:
```diff
@@ -100,23 +100,32 @@ public int available() throws IOException
+
+                // if the chunk was a last chunk (but not an instanceof EOF), then nothing more to do
+                if (chunk.isLast())
+                    return;
```
You should do `chunk = Chunk.next(chunk)` before returning to avoid keeping a ref to a last chunk whose buffer has been released.

gregw replied:
done

ghost commented:
I do not expect this to fix #10513 because that is not related to failing to consuming available content. Quite the opposite, the problem there is that the handling code is blocked trying to consume the content.

Please correct me if I'm wrong on this.

sbordet reviewed (changes requested):
Same comment as @lorban about using `Chunk.next()`.

gregw committed:
Merge branch 'jetty-12.0.x' into fix/10513/ContentSourceInputStreamConsumeAvailable

gregw commented:
@silviobierman perhaps this PR will not fix #10513, as I really can't see what here would affect a http2 only stall.    However, it is some wrongness discovered during investigation of that issue, so should be fixed; should not hurt.   I'll update the discription.

To help with 10513, we need to see some thread dumps of the stall.

gregw committed:
Update from review

use next to avoid reference to non terminal chunk.

gregw committed:
Merge branch 'jetty-12.0.x' into fix/10513/ContentSourceInputStreamConsumeAvailable

gregw committed:
Merge branch 'jetty-12.0.x' into fix/10513/ContentSourceInputStreamConsumeAvailable

sbordet reviewed (changes requested):
I'm not sure about this change.

If I have a 4 GiB file as input, and the code processing that file throws in a try-with-resources, the implicit `close()` is now reading 4 GiB before terminating.

As this is a generic class, we should not assume it's wrapping the `Request` and hence the semantic of `consumeAvailable()` should be done in `close()`.

I agree it's surprising and annoying to have to consume the input until EOF in the application, but this change may possibly hurt more than it's trying to fix.

gregw committed:
updates from review

Only do a single read looking for EOF

gregw committed:
updates from review

Only do a single read looking for EOF
If any content is skipped, then it is an abnormal close.
use Chunk.next in read

gregw commented:
@sbordet @lorban I have a pushed an alternate version that:
 + will always fail the source if any content is skipped by the close - even if that has already been read and not yet consumed.
 + will attempt only 1 read looking for a real EOF or failure (but will fail if there is more content - even if it is the last content).
 + uses Chunk.next in read to handle failures

I think with these changes we are more consistent with regards to how content and EOF are decomposed into chunks:
 + If there is any left over content (1 byte or 4GB) we will fail. 
 + If there is an EOF available after all the consumed content, we will see it.

The implementation is a bit long, but I thought explicitly unrolling the loop of 1 was clearest.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceInputStream.java:
```diff
@@ -98,23 +98,49 @@ public int available() throws IOException
+            // If we read a chunk
+            if (chunk != null)
+            {
+                // Handle a failure as read would
+                if (Content.Chunk.isFailure(chunk))
+                {
+                    Content.Chunk c = chunk;
+                    chunk = Content.Chunk.next(c);
+                    throw IO.rethrow(c.getFailure());
+                }
```
I would not do this, as this is a close() and I don't want an exception in my face if I can't read -- it's a quasi-normal behavior if the other side has already closed as well.
I'd just store the failure (in case of weird extra reads) and return.

sbordet commented:
@gregw I think we should strive for the following: upon close(), if there is a terminal "Transfer-Encoding: chunked" chunk, or zip terminal bytes, or extra newline after JSON, we try to read one chunk only, and if last=true, or an empty last=true chunk, then we silently close with Chunk.EOF.

If there is no chunk we record the failure, if there is a chunk last=false we record the failure.
